spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -168,12 +159,78 @@ public Object fromMessage(Message<?> message, Class<?> targetClass, Object conve
-			return messageBuilderFactory.withPayload(payload).copyHeaders(headers)
+			MimeType messageContentType = MessageConverterUtils.X_JAVA_OBJECT.equals(this.contentType) ?
+					MessageConverterUtils.javaObjectMimeType(payload.getClass()) : this.contentType;
+			return MessageConverterConfigurer.this.messageBuilderFactory.withPayload(payload).copyHeaders(headers)
 					.copyHeaders(Collections.singletonMap(MessageHeaders.CONTENT_TYPE,
 							messageContentType.toString()))
 					.build();
 		}
 	}
+
+	private final class ContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
+
+		private final String contentType;
+
+		private final MimeType mimeType;
+
+		private ContentTypeConvertingInterceptor(String contentType) {
+			this.contentType = contentType;
+			this.mimeType = MessageConverterUtils.getMimeType(contentType);
+		}
+
+		@Override
+		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			Class<?>[] classes =
+					MessageConverterConfigurer.this.compositeMessageConverterFactory.supportedDataTypes(
+							this.mimeType);
+			MessageWrappingMessageConverter messageConverter =
+					new MessageWrappingMessageConverter(
+							MessageConverterConfigurer.this.compositeMessageConverterFactory
+									.getMessageConverterForType(this.mimeType), this.mimeType);
+			for (Class<?> aClass : classes) {
+				if (aClass.isAssignableFrom(message.getPayload().getClass())) {
+					Object contentTypeFromMessage = message.getHeaders().get(MessageHeaders.CONTENT_TYPE);
+					if (contentTypeFromMessage == null) {
+						return MessageConverterConfigurer.this.messageBuilderFactory
+								.fromMessage(message)
+								.setHeader(MessageHeaders.CONTENT_TYPE, this.contentType)
+								.build();
+					}
+					else {
+						return message;
+					}
+				}
+				else {
+					Object converted = messageConverter.fromMessage(message, aClass);
+					if (converted != null) {
+						return (Message<?>) converted;
+					}
+				}
+			}
+			throw new MessageConversionException("Cannot convert " + message + " to " + this.contentType);
+		}
+	}
+
+	private final class PartitioningInterceptor extends ChannelInterceptorAdapter {
+
+		private final BindingProperties bindingProperties;
+
+		private PartitioningInterceptor(BindingProperties bindingProperties) {
+			this.bindingProperties = bindingProperties;
+		}
+
+		@Override
+		public Message<?> preSend(Message<?> message, MessageChannel channel) {
+			EvaluationContext evaluationContext = ExpressionUtils.createStandardEvaluationContext(
+					MessageConverterConfigurer.this.beanFactory);
+			PartitionHandler partitionHandler = new PartitionHandler(MessageConverterConfigurer.this
+					.beanFactory, evaluationContext, null,
+					this.bindingProperties.getProducer());
+			int partition = partitionHandler.determinePartition(message);
+			return new MutableMessageBuilderFactory().fromMessage(message)
+					.setHeader(BinderHeaders.PARTITION_HEADER, partition).build();
```

Comment:
Does this mean, the message will always have the `BinderHeaders.PARTITION_HEADER` with the `partition` value if the producer has `partitioned` consumer and will never have to determine partition at runtime (while that is still possible right before serialization) ?